radix-transaction-scenarios/src/scenarios/all_scenarios.rs

```diff
@@ -1,54 +1,422 @@
+    pub fn scenarios_to_execute(
+        mut self,
+        scenario_requirements: impl IntoIterator<Item = ScenarioRequirements>,
+    ) -> Self {
+        self.scenarios_to_execute = scenario_requirements.into_iter().collect();
+        self
+    }
+
+    /// Controls whether the bootstrap process should be performed or not.
+    pub fn bootstrap(mut self, bootstrap: bool) -> Self {
+        self.bootstrap = bootstrap;
+        self
+    }
+
+    /// Sets the starting nonce for executing scenarios.
+    pub fn starting_nonce(mut self, starting_nonce: u32) -> Self {
+        self.starting_nonce = starting_nonce;
+        self
+    }
+
+    /// Defines how the executor should handle nonces.
+    pub fn nonce_handling(mut self, nonce_handling: NonceHandling) -> Self {
+        self.nonce_handling = nonce_handling;
+        self
+    }
+
+    /// Sets the network definition to use for the scenario execution.
+    pub fn network_definition(mut self, network_definition: NetworkDefinition) -> Self {
+        self.network_definition = network_definition;
+        self
+    }
+
+    /// Sets the callback to call after executing a scenario transaction.
+    pub fn on_transaction_execution<
+        NF1: FnMut(&ScenarioMetadata, &NextTransaction, &TransactionReceiptV1, &D),
+    >(
+        self,
+        callback: NF1,
+    ) -> TransactionScenarioExecutor<D, W, E, NF1, F2> {
+        TransactionScenarioExecutor {
+            /* Environment */
+            database: self.database,
+            scrypto_vm: self.scrypto_vm,
+            native_vm: self.native_vm,
+            /* Execution */
+            scenarios_to_execute: self.scenarios_to_execute,
+            bootstrap: self.bootstrap,
+            starting_nonce: self.starting_nonce,
+            nonce_handling: self.nonce_handling,
+            network_definition: self.network_definition,
+            /* Callbacks */
+            on_transaction_execution: callback,
+            on_scenario_start: self.on_scenario_start,
+        }
+    }
+
+    /// A callback that is called when a new scenario is started.
+    pub fn on_scenario_start<NF2: FnMut(&ScenarioMetadata)>(
+        self,
+        callback: NF2,
+    ) -> TransactionScenarioExecutor<D, W, E, F1, NF2> {
+        TransactionScenarioExecutor {
+            /* Environment */
+            database: self.database,
+            scrypto_vm: self.scrypto_vm,
+            native_vm: self.native_vm,
+            /* Execution */
+            scenarios_to_execute: self.scenarios_to_execute,
+            bootstrap: self.bootstrap,
+            starting_nonce: self.starting_nonce,
+            nonce_handling: self.nonce_handling,
+            network_definition: self.network_definition,
+            /* Callbacks */
+            on_transaction_execution: self.on_transaction_execution,
+            on_scenario_start: callback,
+        }
+    }
+
+    pub fn execute(mut self) -> Result<CompletedExecutionReceipt<D>, ScenarioExecutorError> {
```

Comment:
I find it a little confusing that this just says `execute` when it's called, and not `execute_all` or something.

Actually, I think a much clearer API would be: `execute(self, scenario_filter: ScenarioFilter)`

Where we have e.g.
```rust
enum ScenarioFilter {
    ExactScenarios([String]),
    AllScenariosValidAtCurrentProtocolVersion, // Would need to save this on the LedgerSimulator I guess?
    AllScenariosValidAsOf(ProtocolUpdate),
    AllScenariosFirstValidAt(ProtocolUpdate),
}
```

Reply:
Filtering sounds good to me especially with the refactor I made to the executor and how it works I think that filtering would fit in nicely. although I don't think that it should be a parameter passed to `execute` but rather `.filter` and `.remove_filter` methods on the builder.

The `AllScenariosValidAtCurrentProtocolVersion` is problematic since the executor can't tell what the current protocol version is from the substate store alone and so I can't see how we would find this information, unless we make it part of the "store protocol config in substates" ticket.

Reply:
> although I don't think that it should be a parameter passed to execute but rather .filter and .remove_filter methods on the builder.

I think the default behaviour is quite unclear with respect to which scenarios actually execute. But if you feel strongly about this, can we at least rename `execute` to `execute_all_matching` or something?

Reply:
I think `execute` can be renamed to be more descriptive while still allowing for a simple interface, I propose:

* `execute_all` - Runs all of the scenarios without any filter.
* `execute_all_matching` - Acceps a filter and executes all scenarios that match/pass through some filter.

How do you feel about that?

Reply:
Sure, we could do that. I feel `execute(ScenarioFilter::All)` is nicer though :). On reflection, I don't mind too much. We can keep it as `execute` for now and I can stick in a PR with a few small tweaks and get your reflections on them?